transformer_engine/pytorch/fp8.py

```diff
@@ -1112,34 +1112,15 @@ def __init__(
             device = torch.device("cuda")
         self.device = device
 
-        if getattr(recipe, "qfactories", None) is None:
-            raise ValueError("CustomRecipe requires `qfactories` with factories.")
+        if getattr(recipe, "qfactory", None) is None:
+            raise ValueError("CustomRecipe requires `qfactory`.")
 
     def make_quantizers(self) -> list:
-        qfactories = self.recipe.qfactories
-        # Build a canonical, per-mode pool of factories and cycle it to match num_quantizers.
-        if self.mode == "forward":
-            pool = [
-                qfactories.input_factory,
-                qfactories.weight_factory,
-                qfactories.output_factory,
-            ]
-        else:
-            pool = [
-                qfactories.grad_output_factory,
-                qfactories.grad_input_factory,
-            ]
-
-        # Filter out None entries; allow partial factories and reuse as needed.
-        pool = [factory for factory in pool if factory is not None]
-        if len(pool) == 0:
-            raise ValueError(
-                "CustomRecipe requires at least one factory for the requested mode; "
-                f"got none for mode={self.mode}."
-            )
-
+        qfactory = self.recipe.qfactory
         out = []
         for i in range(self.num_quantizers):
-            factory = pool[i % len(pool)]
-            out.append(factory())
+            # TODO(negvet): Support other operations.
+            role = self.recipe.resolve_role("linear", self.mode, i)
+            quantizer = qfactory(role)
```

Comment:
Separating out different roles is a nice way to handle different operations. Could we go even further by removing `resolve_role` and just passing everything into the factory function?

Pros:
- The factory functions have complete flexibility over what quantizers they produce. They can add new operations/roles without updating TE.
- A naive quantization scheme might just return the same quantizer for everything, in which case resolving the role is unnecessary.
- The role names are linear-specific. I'd expect convolution to also have an input, weights, and output.

Cons:
- The factory functions need to have some awareness of the quantizer order.
- Role names are more descriptive than indices.

Reply:
I think I have an even nicer long-term solution, although we don't need to implement it quite yet. Instead of resolving roles inside `CustomRecipeState`, we should change `RecipeState.make_quantizers` so that it can take in a list of roles. Each operation can then choose whatever order it likes and not be bound to a cyclic order.

For now, let's settle on an external API of just `qfactory(role)`. If we're keeping the roles, the kwargs are not adding much value.

Reply:
This is roughly what I proposed as well, if you check TODO in make_quantizers() in https://github.com/NVIDIA/TransformerEngine/pull/2039/commits/7d6fef5c1ce028a9795f879c1dca1b4d5f12edda#diff-77a87478142dbcec64bbe6fc9f402be4c4c9aa21e375cca6d352545fa8e87166R1140